api/src/main/java/com/nicknathanjustin/streamercontracts/votes/VoteServiceImpl.java

```diff
@@ -0,0 +1,130 @@
+        if (proposerMarkedContractCompleted(contractModel)) {
+            //Proposer has marked contract as completed. Release funds without reading Streamer's vote.
+            return true;
+        } else if (streamerMarkedContractFailed(contractModel)) {
+            //Streamer has marked contract as failed. Return funds without reading Proposers's vote.
+            return false;
+        } else if(proposerVote == null && streamerVote == null) {
+            //Neither proposer or streamer voted. Release funds if contract was accepted. Return funds otherwise.
+            return contractModel.isAccepted();
+        } else if(proposerVote != null) {
+            //Streamer did not vote. Take Proposer's vote as the final outcome
+            return proposerVote.isViewerFlaggedComplete();
+        } else if(streamerVote != null){
+            //Proposer did not vote. Take Streamer's vote as the final outcome
+            return streamerVote.isViewerFlaggedComplete();
+        }
+        //Default case is to consider the contract completed and release money to the streamer.
+        return true;
+    }
+
+    private boolean proposerMarkedContractCompleted(@NonNull final ContractModel contractModel) {
+        final VoteModel proposerVote = voteModelRepository.findByContractIdAndVoterId(contractModel.getId(), contractModel.getProposer().getId());
+        return proposerVote != null && proposerVote.isViewerFlaggedComplete();
+    }
+
+    private boolean streamerMarkedContractFailed(@NonNull final ContractModel contractModel) {
+        final VoteModel streamerVote = voteModelRepository.findByContractIdAndVoterId(contractModel.getId(), contractModel.getStreamer().getId());
+        return streamerVote != null && !streamerVote.isViewerFlaggedComplete();
+    }
+
+    private boolean userCanVoteOnContract(@NonNull final UserModel voter, @NonNull final ContractModel contractModel) {
+        validateContract(contractModel);
+
+        final UUID contractId = contractModel.getId();
+
+        final UUID proposerId = contractModel.getProposer().getId();
+        final boolean isVoterContractProposer = voter.getId().equals(proposerId);
+        final boolean proposerHasVoted = (voteModelRepository.findByContractIdAndVoterId(contractId, proposerId) != null);
+
+        final UUID streamerId = contractModel.getStreamer().getId();
+        final boolean isVoterContractStreamer = voter.getId().equals(streamerId);
+        final boolean streamerHasVoted = (voteModelRepository.findByContractIdAndVoterId(contractId, streamerId) == null);
+
+        final boolean isUserAllowedToVote = isVoterContractProposer || isVoterContractStreamer;
+        if (!isUserAllowedToVote) {
+            log.warn("Unauthorized vote detected. User: {} attempted to vote on contractId: {} without having permissions to do so",
+                    voter.getId(),
+                    contractModel.getId());
```

Comment:
Just return false right away here? We don't need to check if proposer/streamer has already voted if the user is not allowed to vote

Reply:
Yuppers. Good callout. fixed.